modules/red/red.js

```diff
@@ -0,0 +1,128 @@
+                this.frameBuffer[i] = currentBuffer[i + this.targetRedundancy - targetRedundancy];
+            }
+        }
+        this.targetRedundancy = targetRedundancy;
+    }
+
+    /**
+     * Set the "inner opus payload type". This is typically our RED payload type that we tell
+     * the other side as our opus payload type. Can be queried from the sender using getParameters()
+     * after setting the answer.
+     * @param {Number} payloadType the payload type to use for opus.
+     */
+    setOpusPayloadType(payloadType) {
+        this.payloadType = payloadType;
+    }
+
+    /**
+     * This is the actual transform to add redundancy to a raw opus frame.
+     * @param {RTCEncodedAudioFrame} encodedFrame - Encoded audio frame.
+     * @param {TransformStreamDefaultController} controller - TransportStreamController.
+     */
+    addRedundancy(encodedFrame, controller) {
+        // TODO: should this ensure encodedFrame.type being not set and
+        // encodedFrame.getMetadata().payloadType being the same as before?
+        /*
+         * From https://datatracker.ietf.org/doc/html/rfc2198#section-3:
+         0                   1                    2                   3
+         0 1 2 3 4 5 6 7 8 9 0 1 2 3  4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+         +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+         |F|   block PT  |  timestamp offset         |   block length    |
+         +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+         0 1 2 3 4 5 6 7
+         +-+-+-+-+-+-+-+-+
+         |0|   Block PT  |
+         +-+-+-+-+-+-+-+-+
+         */
+        const data = new Uint8Array(encodedFrame.data);
+
+        const newFrame = data.slice(0);
+
+        newFrame.timestamp = encodedFrame.timestamp;
+
+        let allFrames = this.frameBuffer.filter(x => Boolean(x)).concat(newFrame);
+
+        // TODO: determine how much we can fit into the available size (which we need to assume as 1190 bytes or so)
+        let needLength = 1 + newFrame.length;
+
+        for (let i = allFrames.length - 2; i >= 0; i--) {
+            const frame = allFrames[i];
+
+
+            // TODO: timestamp wraparound?
+            if ((allFrames[i + 1].timestamp - frame.timestamp + MAX_TIMESTAMP) % MAX_TIMESTAMP >= 16384) {
+                allFrames = allFrames.slice(i + 1);
+                break;
+            }
+            needLength += 4 + frame.length;
+        }
+
+        const newData = new Uint8Array(needLength);
+        const newView = new DataView(newData.buffer);
+
+        // Construct the header.
+        let frameOffset = 0;
+
+        for (let i = 0; i < allFrames.length - 1; i++) {
+            const frame = allFrames[i];
+
+            // Ensure correct behaviour on wraparound.
+            const tOffset = (encodedFrame.timestamp - frame.timestamp + MAX_TIMESTAMP) % MAX_TIMESTAMP;
+
+            newView.setUint8(frameOffset, this.payloadType | 0x80); // eslint-disable-line no-bitwise
```

Comment:
it throws and throwing inside this callback is bad... but it is not like you have any alternative to throwing in that case because it is a programmer error

Reply:
Awesome. Wrong API sequencing. Well not much we can do about it it seems.